src/swarm-lang/Swarm/Language/Types.hs

```diff
@@ -816,8 +824,50 @@ data TydefInfo = TydefInfo
 makeLenses ''TydefInfo
 
 -- | A @TDCtx@ is a mapping from user-defined type constructor names
---   to their definitions and arities/kinds.
-type TDCtx = Ctx TydefInfo
+--   to their definitions and arities/kinds.  It also stores the
+--   latest version of each name (for any names with more than one
+--   version), so we can tell when a type definition has been
+--   shadowed.
+data TDCtx = TDCtx {getTDCtx :: Ctx TydefInfo, getTDVersions :: Map Text Int}
```

Comment:
Monoid Map for Sum might be nice to eliminate the fromMaybe.

Reply:
Hmm, that sounds like a nice idea in theory, but I just tried it and ran into an issue with missing `Hashable` and `ToJSON` instances for `MonoidMap`.  I suppose we could add some orphan instances that work via conversion to `Map`...

Reply:
It's not an orphan if you declare it for our TDCtx type :)

Reply:
Good point!